### PR TITLE
Book: backend extension + custom wgpu kernel

### PR DESCRIPTION
**Pull request by nathanielsimard**
During the writing of this guide, I had to change the visibility of some functions and I also fixed some typos.

Overall this is a pretty big PR that document how to create your own backend extension with a fused WGPU kernel as an example.

### CODE REVIEW

**Comment by antimora** [comment]
It's interesting to see how book/guide forces one to think/revisit from the user's perspective, consequently making you change some APIs.

File: examples/custom-wgpu-kernel/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+mod backward;
+mod forward;
+
+use burn::tensor::{activation, Tensor};
+
+/// We create a type alias to avoid the ugly disambiguation with associative types.
+pub type FloatTensor<B, const D: usize> = <B as burn::tensor::backend::Backend>::TensorPrimitive<D>;
```
Comment (louisfd) [inline examples/custom-wgpu-kernel/src/lib.rs:7]: This makes me think, the concept of TensorPrimitive is not really explained in the previous sections

File: burn-book/src/advanced/backend-extension/custom-wgpu-kernel.md [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+        let lhs_index = row * K + k;
+        let rhs_index = k * n_cols + col;
+
+        sum += lhs[offset_lhs + lhs_index] * rhs[offset_rhs + rhs_index];
+    }
+
+    let output_index = row * n_cols + col;
+    let index = offset_output + output_index;
+
+    output[index] = max(sum + bias[index], 0.0);
```
Comment (louisfd) [inline burn-book/src/advanced/backend-extension/custom-wgpu-kernel.md:148]: Maybe a comment to emphasize the fact that the add and relu are done here

File: burn-book/src/advanced/backend-extension/custom-wgpu-kernel.md [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        let blocks_needed_in_x = f32::ceil(num_rows as f32 / workgroup_size_x as f32) as u32;
+        let blocks_needed_in_y = f32::ceil(num_cols as f32 / workgroup_size_y as f32) as u32;
+
+        // Compile the kernel or use the cache based on the template id.
+        let kernel = lhs.context.compile_dynamic(FusedMatmulAddRelu::<F>::new(
+            workgroup_size_x,
+            workgroup_size_y,
+        ));
+
+        // Build info buffer with tensor information.
```
Comment (louisfd) [inline burn-book/src/advanced/backend-extension/custom-wgpu-kernel.md:228]: Add details on what this means

File: burn-book/src/advanced/backend-extension/custom-wgpu-kernel.md [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+        let info = build_info(&[&lhs, &rhs, &output]);
+        let info_buffer = lhs
+            .context
+            .create_buffer_with_data(bytemuck::cast_slice(&info));
+
+        // Calculate launch information.
+        let mut num_iter = 1;
+        for i in 0..D - 2 {
```
Comment (louisfd) [inline burn-book/src/advanced/backend-extension/custom-wgpu-kernel.md:236]: If we're gonna have a for loop on the batches, might as well compute the output shape at the same moment and save us from having to explain shape_out

File: burn-book/src/advanced/backend-extension/custom-wgpu-kernel.md [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+                if let Some(node) = node_lhs {
+                    grads.register::<B, D>(node, grad_lhs);
+                }
+                if let Some(node) = node_rhs {
+                    grads.register::<B, D>(node, grad_rhs);
+                }
+            }
+        }
+
+        // Prepare a stateful operation with each variable node and corresponding graph.
```
Comment (louisfd) [inline burn-book/src/advanced/backend-extension/custom-wgpu-kernel.md:351]: Unclear to me what preparation does

**Comment by louisfd** [comment]
I updated the branch with corrections regarding my previous comments. 
There is still the part about preparation in the backward pass that remains unclear to me, then again this should probably be explained earlier on in the book.

**Comment by louisfd** [comment]
@nathanielsimard Can you review the last commit please? I wrote an entire paragraph which I think is clearer but I want to make sure I did not say anything untrue

**Review by louisfd — APPROVED** [review]

**Pull request merged**